### PR TITLE
fix(server): fix getex command with persist can get time overflow

### PR DESCRIPTION
**Pull request by boazsade**
Signed-off-by: Boaz Sade <boaz@dragonflydb.io>

This fix issue #705 
GETEX command with PERSIST will result in integer overflow since in this case, the expiration time cannot be calculated.
The fix is, in case we have persist enable, don't try to calculate the persist time, just set as persist value.

### CODE REVIEW

**Comment by dranikpg** [comment]
Interestingly, I found the same issue today and fixed it, but not as well structured as with your solution 🙂 

https://github.com/dragonflydb/dragonfly/pull/699/files#diff-883d71d37722157c22061e0b592d4784172d9e6f50a8cbf0c76f349e483418e6

**Review by dranikpg — DISMISSED** [review]

File: src/server/db_slice.cc [hunk 1]
```diff
@@ -597,23 +597,29 @@ pair<int64_t, int64_t> DbSlice::ExpireParams::Calculate(int64_t now_ms) const {
 
 OpResult<int64_t> DbSlice::UpdateExpire(const Context& cntx, PrimeIterator prime_it,
                                         ExpireIterator expire_it, const ExpireParams& params) {
+  constexpr uint64_t kPersistValue = 0;
   DCHECK(params.IsDefined());
   DCHECK(IsValid(prime_it));
+  // If this need to persist, then only set persist value and return
+  if (params.persist) {
+    UpdateExpire(cntx.db_index, prime_it, kPersistValue);
```
Comment (romange) [inline src/server/db_slice.cc:605]: I would like to deprecate this UpdateExpire function, specifically please use here RemoveExpire.

File: src/server/db_slice.cc [hunk 2]
```diff
@@ -597,23 +597,29 @@ pair<int64_t, int64_t> DbSlice::ExpireParams::Calculate(int64_t now_ms) const {
...
+  if (rel_msec <= 0) {  // implicit - don't persist
     CHECK(Del(cntx.db_index, prime_it));
     return -1;
   } else if (IsValid(expire_it)) {
     expire_it->second = FromAbsoluteTime(abs_msec);
     return abs_msec;
   } else {
-    UpdateExpire(cntx.db_index, prime_it, params.persist ? 0 : abs_msec);
-    return params.persist ? 0 : abs_msec;
+    UpdateExpire(cntx.db_index, prime_it, abs_msec);
```
Comment (romange) [inline src/server/db_slice.cc:621]: If I am not mistaken, here it should be `AddExpire`.
And if `UpdateExpire(DbIndex db_ind, PrimeIterator it, uint64_t at)` is not called anymore please remove it as well.

**Comment by romange** [comment]
Boaz, Since Vlad already merged his code, please remove Vlad's check :)
You could pass a dcheck there that persist is false instead.

File: src/server/db_slice.cc (hunk 1 above)
Comment (boazsade) [reply src/server/db_slice.cc:605]: OK

File: src/server/db_slice.cc (hunk 2 above)
Comment (boazsade) [reply src/server/db_slice.cc]: @romange - please let me know if these changes are not what you have in mind in term of removing "UpdateExpire" function

File: src/server/generic_family.cc [hunk 3]
```diff
@@ -397,7 +397,9 @@ OpStatus Renamer::UpdateDest(Transaction* t, EngineShard* es) {
         dest_it->second = std::move(pv_);
       }
       dest_it->second.SetExpire(has_expire);  // preserve expire flag.
-      db_slice.UpdateExpire(t->GetDbIndex(), dest_it, src_res_.expire_ts);
+      if (!has_expire && src_res_.expire_ts) {
+        db_slice.AddExpire(t->GetDbIndex(), dest_it, src_res_.expire_ts);
```
Comment (romange) [inline src/server/generic_family.cc:401]: i suspect it's wrong. please revert this change (unfortunately)

File: src/server/generic_family.cc [hunk 4]
```diff
@@ -1352,7 +1353,9 @@ OpResult<void> GenericFamily::OpRen(const OpArgs& op_args, string_view from_key,
     // It is guaranteed that UpdateExpire() call does not erase the element because then
     // from_it would be invalid. Therefore, UpdateExpire does not invalidate any iterators,
     // therefore we can delete 'from_it'.
-    db_slice.UpdateExpire(op_args.db_cntx.db_index, to_it, exp_ts);
+    if (!to_it->second.HasExpire() && exp_ts) {
+      db_slice.AddExpire(op_args.db_cntx.db_index, to_it, exp_ts);
```
Comment (romange) [inline src/server/generic_family.cc:1357]: please revert this as well

File: src/server/string_family.cc [hunk 5]
```diff
@@ -402,9 +402,9 @@ OpStatus SetCmd::Set(const SetParams& params, string_view key, string_view value
   it->second = std::move(tvalue);
   db_slice.PostUpdate(op_args_.db_cntx.db_index, it, key, false);
 
-  if (params.expire_after_ms) {
-    db_slice.UpdateExpire(op_args_.db_cntx.db_index, it,
-                          params.expire_after_ms + op_args_.db_cntx.time_now_ms);
+  if (params.expire_after_ms && !it->second.HasExpire()) {
```
Comment (romange) [inline src/server/string_family.cc:405]: ` && !it->second.HasExpire()` no needed, it's a a new value.

File: src/server/generic_family.cc (hunk 4 above)
Comment (boazsade) [reply src/server/generic_family.cc:1357]: sure

File: src/server/generic_family.cc (hunk 3 above)
Comment (boazsade) [reply src/server/generic_family.cc:401]: sure

File: src/server/string_family.cc (hunk 5 above)
Comment (boazsade) [reply src/server/string_family.cc:405]: ok

**Review by romange — APPROVED** [review]
